### ISSUE TITLE
Out-RsCatalogItem not generating correct XML for data sources

### ISSUE THREAD

**Issue by sqlchicken** [Microsoft/ReportingServicesTools#9]
When using Out-RsCatalogItem to export a data source, the API is not returning expected/correct XML definition. You cannot deploy an exported dataset back to Reporting Services without the proper formatting. Reporting Services does not see it as a valid data source.

Below are comparisons of RDS definitions:

Exported RDS via SSRS API (GetItemDefinition):
<DataSourceDefinition>
	<Extension>SQL</Extension>
	<ConnectString>Data Source=(local);Initial Catalog=AdventureWorks2014</ConnectString>
	<UseOriginalConnectString>False</UseOriginalConnectString>
	<OriginalConnectStringExpressionBased>False</OriginalConnectStringExpressionBased>
	<CredentialRetrieval>Store</CredentialRetrieval>
	<WindowsCredentials>True</WindowsCredentials>
	<ImpersonateUser>False</ImpersonateUser>
	<UserName>domain\user</UserName>
	<Enabled>True</Enabled>
</DataSourceDefinition>

Definition of shared data source as created in Visual Studio:
<?xml version="1.0" encoding="utf-8"?>
<RptDataSource xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" Name="dsrcAW2016">
	<ConnectionProperties>
		<Extension>SQL</Extension>
		<ConnectString>Data Source=(local)\elektra;Initial Catalog=AdventureworksDW2016CTP3</ConnectString>
		<IntegratedSecurity>true</IntegratedSecurity>
	</ConnectionProperties>
	<DataSourceID>dc94beb9-b13c-4219-ae9e-ea5033017691</DataSourceID>
</RptDataSource>

**Comment by sqlchicken** [comment]
This looks like a promising place to derive solution code from: [MSSQL Tips article](https://www.mssqltips.com/sqlservertip/4429/sql-server-reporting-services-data-source-deployment-automation-with-powershell/)

**Comment by propellor** [comment]
I tried using the output from the referenced Gallery Solution, but ended up with the same error.  I ran that script and then replaced the ReportingServicesTools data source xml content.  Not really sure what the issue is.

[error.txt](https://github.com/Microsoft/ReportingServicesTools/files/702876/error.txt)

**Comment by sqlchicken** [comment]
Ultimately the issue is that Reporting Services isn't exporting the proper XML format. I compared the XML of a valid data source versus one exported. Here's one exported via the gallery solution method (exported rds doesn't work):

Good example:
<?xml version="1.0" encoding="utf-8"?>
<RptDataSource xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" Name="dsAdventureWorks">
  <ConnectionProperties>
    <Extension>SQL</Extension>
    <ConnectString>Data Source=(local);Initial Catalog=AdventureWorks2014</ConnectString>
    <IntegratedSecurity>true</IntegratedSecurity>
  </ConnectionProperties>
  <DataSourceID>3d7902c5-9aba-4cb0-b59d-d2ae2c205721</DataSourceID>
</RptDataSource>


Bad export:
<?xml version="1.0" encoding="utf-8"?>
<DataSourceDefinition xmlns="http://schemas.microsoft.com/sqlserver/reporting/2006/03/reportdatasource">
  <Extension>SQL</Extension>
  <ConnectString>Data Source=.\elektra;Initial Catalog=ReportServer;</ConnectString>
  <CredentialRetrieval>Store</CredentialRetrieval>
  <WindowsCredentials>False</WindowsCredentials>
  <ImpersonateUser>False</ImpersonateUser>
  <Enabled>True</Enabled>
</DataSourceDefinition>

**Comment by propellor** [comment]
Using that both of those as input doesn't seem to work for me either the same error or one complaining about invalid xml schema - maybe something to do with my environment.  The method from this site _does_ work:  http://bretstateham.com/extracting-ssrs-report-rdl-xml-from-the-reportserver-database/

See contentvarchar column in this query.  I modified it to only show datasources.

```

--The first CTE gets the content as a varbinary(max)
--as well as the other important columns for all reports,
--data sources and shared datasets.
WITH ItemContentBinaries AS
(
  SELECT
     ItemID,Name,[Type]
    ,CASE Type
       WHEN 2 THEN 'Report'
       WHEN 5 THEN 'Data Source'
       WHEN 7 THEN 'Report Part'
       WHEN 8 THEN 'Shared Dataset'
       ELSE 'Other'
     END AS TypeDescription
    ,CONVERT(varbinary(max),Content) AS Content
  FROM ReportServer.dbo.Catalog
  WHERE Type IN (5,8)
),
--The second CTE strips off the BOM if it exists...
ItemContentNoBOM AS
(
  SELECT
     ItemID,Name,[Type],TypeDescription
    ,CASE
       WHEN LEFT(Content,3) = 0xEFBBBF
         THEN CONVERT(varbinary(max),SUBSTRING(Content,4,LEN(Content)))
       ELSE
         Content
     END AS Content
  FROM ItemContentBinaries
)
--The outer query gets the content in its varbinary, varchar and xml representations...
SELECT
   ItemID,Name,[Type],TypeDescription
  ,Content --varbinary
  ,CONVERT(varchar(max),Content) AS ContentVarchar --varchar
  ,CONVERT(xml,Content) AS ContentXML --xml
FROM ItemContentNoBOM
/*WHERE name = 'dba_rep'*/
```

**Comment by parthsha** [comment]
@sqlchicken and @propellor I have made changes to Write-RsCatalogItem such that it is now able to create data source based on the DataSourceDefinition xml. Can you please try the new version of this function and tell me what do you think about it? I was wondering whether this change would be sufficient to close this issue.

**Comment by propellor** [comment]
@parthsha, thanks for that update.  I have tested again against SSRS 2012 and while there is improvement in the data source, the reports are broken.

1. The data source is now created, however, only the name remains.  Server/Provider info is lost.  I expect credentials to be reset, but usually that information should transfer.  

2.  Report files are no longer XML text files.  They show as some type of Asian character set in text editors, perhaps Chinese. 믯㲿砿汭

Thanks again for working on this.

**Comment by parthsha** [comment]
When you say Server/Provider Info is lost, what do you mean? I see the connection string (see <ConnectString> and the type of data source it is <Extension>. Is there something else you are looking for? The credentials will be reset but that is more for security purposes than functional.

**Comment by propellor** [comment]
Yes those two properties you mention do not make it into the uploaded data source.  All I get is a data source with the correct name.

**Comment by parthsha** [comment]
@propellor I made a silly error. I am sorry about that. Could you please try again?

**Comment by sqlchicken** [comment]
It’s still not working properly. Still not seeing valid definition for the exported items.

**Comment by parthsha** [comment]
@sqlchicken the XML for data source is not what you have mentioned above. I was trying to solve the issue of being able to upload the data source that was downloaded by the PowerShell scripts. I will try looking into getting the correct XML as soon as possible.

**Comment by DMiradakis** [comment]
This is still broken as of 2021-01-04.
